### PR TITLE
Move cooking types to Declarations from Opaqueproof

### PR DESCRIPTION
**Pull request by SkySkimmer**
This is the highest they can go and may fit better (opaqueproof doesn't care what the cooking_info type contains).

Overlay: https://github.com/MetaCoq/metacoq/pull/560

### CODE REVIEW

File: kernel/declarations.ml [hunk 1]
```diff
@@ -100,9 +100,17 @@ type typing_flags = {
...
+
+type cooking_info = {
+  modlist : work_list;
+  abstract : Constr.named_context * Univ.Instance.t * Univ.AUContext.t;
+}
+
 (* some contraints are in constant_constraints, some other may be in
  * the OpaqueDef *)
-type 'opaque constant_body = {
+type 'opaque constant_body_g = {
```
Comment (ppedrot) [inline kernel/declarations.ml:123]: I am not sure I like this naming scheme. We already use it for other purposes, and in the kernel we rather prefix parameteric types with `p`.

**Comment by ppedrot** [comment]
Note that I am currently working again on a patch to remove the opaque table and the future proofs from the kernel. With it there is no need to track the type of the opaque data since the table essentially becomes an integer tracking the number of allocated opaques. That would allow pushing this stuff even higher, e.g. in Safe_typing.

File: kernel/section.mli [hunk 2]
```diff
@@ -68,27 +68,16 @@ val all_poly_univs : 'a t -> Univ.Level.t array
    constraints about monomorphic universes, which prevent declaring
    monomorphic globals. *)
 
-type abstr_info = private {
```
Comment (ppedrot) [inline kernel/section.mli:71]: I've realized that we lost the private flag by setting this in Declarations. This was ensuring that the only way to build this data was in Section, but I don't know how much it matters.

Comment (SkySkimmer) [reply kernel/section.mli:71]: Considering it either was immediately destructed in safe_typing or went outside the kernel I don't think it brought anything.

Comment (ppedrot) [reply kernel/section.mli:71]: Fair enough.

File: kernel/declarations.ml (hunk 1 above)
Comment (ppedrot) [reply kernel/declarations.ml:123]: @SkySkimmer this is the only issue that prevents me from merging this PR, I believe.

Comment (SkySkimmer) [reply kernel/declarations.ml:123]: >We already use it for other purposes

Like what?

Isn't the `p` stuff about universes? The only examples I remember are things like `pconstant` (not parametric) or `puniverses`.

Comment (ppedrot) [reply kernel/declarations.ml:123]: I think that `p` stands both for parametric or polymorphic. We have it in constr also for generic terms. But in any case, the `_g` form is used in the upper layers for CAst-wrapped ASTs, so it shouldn't be used in the kernel for something unrelated.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
Not sure whether I should merge since we're in a quantum freezing state, I'll wait a bit to see if things sort themselves out alone.

**Pull request merged**